rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+            if (VersionedDiagnostics.setIfNewer(box, newer)) {
+                sendDiagnostics();
+            }
+            return newer;
         }
 
-        public void invalidate(boolean isClosing) {
+        private void invalidate(ParametricSummaryBridge summary, boolean isClosing) {
             summary.invalidate(isClosing);
         }
 
-        public void calculate() {
-            summary.calculateSummary();
-            summary.getMessages().thenAccept(this::reportTypeCheckerMessages);
+        public void invalidateAnalyzer(boolean isClosing) {
+            invalidate(summaryAnalyzer, isClosing);
+        }
+
+        public void invalidateBuilder(boolean isClosing) {
+            invalidate(summaryBuilder, isClosing);
+        }
+
+        /**
+         * @param version the version of the file for which summary calculation
+         * is currently requested
+         * @param latestVersion the last version of the file for which summary
+         * calculation was previously requested
+         * @param delay the duration after which the current request for summary
+         * calculation will be granted, unless another request is made in the
+         * meantime (in which case the current request is abandoned)
+         * @param calculation the actual summary calculation
+         * @return a future that provides optional diagnostics that result from
+         * the summary calculation
+         */
+        private CompletableFuture<Optional<VersionedDiagnostics>> calculate(
+                int version, AtomicInteger latestVersion, Duration delay,
+                Supplier<CompletableFuture<Optional<VersionedDiagnostics>>> calculation) {
+
+            latestVersion.set(version);
+            var delayed = CompletableFuture.delayedExecutor(delay.toMillis(), TimeUnit.MILLISECONDS, exec);
+            return CompletableFuture.supplyAsync(() -> {
+                // If no new call to `calculate` has been made after `delay` has
+                // passed (i.e., `lastVersion` hasn't changed in the meantime),
+                // then run the calculation. Else, abandon this calculation.
+                if (latestVersion.get() == version) {
+                    return calculation.get();
+                } else {
+                    return CompletableFuture.completedFuture(Optional.<VersionedDiagnostics>empty());
+                }
+            }, delayed).thenCompose(Function.identity());
+        }
+
+        public void calculateAnalyzer(int version, Duration delay) {
+            calculateAnalyzerLatestResult = calculate(version, latestVersionCalculateAnalyzer, delay, () -> {
+                var analysis = summaryAnalyzer.calculateSummary();
+                var analysisTree = analysis.input;
+                var analysisMessages = analysis.output.get().thenApply(Supplier::get);
+                return analysisTree.thenCombine(analysisMessages, (tree, messages) -> {
```

Comment:
not a big fan of the chaining that we need to do, do we need to be responsible to merge the parse tree messages with those from the analyzer? can we not disconnect that?

Reply:
Not a big fan either, but we need to store tree+messages together, so one way or the other, the results of the two futures (`analysisTree` and `analysisMessages`) need to be combined.

An alternative is to combine tree+messages in `calculateSummary` of `ParametricSummaryBridge`, but I feel we'd only be moving complexity from one place to another one, without reducing it (and I wanted to minimize the number of changes to `ParametricSummaryBridge`).

Reply:
Okay, that makes sense, I will say that if ParametricSummaryBridge is the more natural place where this should land, please park it there ;)